### ISSUE TITLE
Size difference between input and target in DiceLoss()?

### ISSUE THREAD

**Issue by farazkhan86** [wolny/pytorch-3dunet#5]
The output from my 3D network has two output channels and a batch size of 1. So for a volume of size ``32*64*64`` my model produces an output of size ``[1, 2, 32, 64, 64]``. My target or label volume has a size of ``[1, 32, 64, 64]``, as it is a binary volume having a single channel. This worked fine with ``torch.nn.CrossEntropyLoss()`` but is not working with the ``DiceLoss()`` here.

I even tried adding a dimension to my label volume by using ``torch.unsqueeze()`` but even then the assert check set in ``DiceLoss()`` fails

`assert input.size() == target.size()`

In what format is the ``DiceLoss()`` expecting the inputs to be? 

Some help would be greatly appreciated.

Many thanks

**Comment by wolny** [comment]
Hi @farazkhan86 thanks for pointing this out. In order to use Binary Cross Entropy or the Dice Loss, the data need to have a specific format:
1. The target (label) volume has to be binary (labels `0-1`) and 4D - i.e. each binary 3D mask in separate channel. If you want to predict just a single channel then `[1, 32, 64, 64]` is a correct size.
2. You need to make sure that the number of the output channels matches the number of channels in the target (label) volume. Your model produces `[1, 2, 32, 64, 64]` outputs, which means that you're trying to train with `--in-channels 2`, use `--in-channels 1` instead.

I've also updated the README to make it clear: https://github.com/wolny/pytorch-3dunet#important

**Comment by wolny** [comment]
JTLYK: I've changed the code slightly so that the Dice Loss can now be used with both 3D and 4D target volume. Pull the last master and your training should work without any modifications to your volume.

**Comment by farazkhan86** [comment]
That is excellent. Many thanks @wolny 
I'll give it a try
